p2p/server.py

```diff
@@ -82,136 +85,187 @@ def __init__(self,
+                self.bootstrap_nodes = ROPSTEN_BOOTNODES
             else:
                 self.logger.warn("No bootstrap nodes for network id: {}".format(network_id))
-                bootstrap_nodes = []
-        self.discovery = DiscoveryProtocol(
-            self.privkey, self.address, bootstrap_nodes=bootstrap_nodes)
-        self.peer_pool = peer_pool_class(
-            peer_class,
-            self.headerdb,
-            self.network_id,
-            self.privkey,
-            self.discovery,
-            min_peers=min_peers,
-        )
-        self.syncer = FullNodeSyncer(chain, chaindb, base_db, self.peer_pool, self.cancel_token)
+                self.bootstrap_nodes = []
 
     async def refresh_nat_portmap(self) -> None:
         """Run an infinite loop refreshing our NAT port mapping.
 
         On every iteration we configure the port mapping with a lifetime of 30 minutes and then
         sleep for that long as well.
         """
-        lifetime = 30 * 60
         while not self.is_finished:
-            self.logger.info("Setting up NAT portmap...")
-            # This is experimental and it's OK if it fails, hence the bare except.
             try:
-                await self._add_nat_portmap(lifetime)
-            except Exception as e:
-                if (isinstance(e, upnpclient.soap.SOAPError) and
-                        e.args == (718, 'ConflictInMappingEntry')):
-                    # An entry already exists with the parameters we specified. Maybe the router
-                    # didn't clean it up after it expired or it has been configured by other piece
-                    # of software, either way we should not override it.
-                    # https://tools.ietf.org/id/draft-ietf-pcp-upnp-igd-interworking-07.html#errors
-                    self.logger.info("NAT port mapping already configured, not overriding it")
-                else:
-                    self.logger.exception("Failed to setup NAT portmap")
-
-            try:
-                await wait_with_token(asyncio.sleep(lifetime), token=self.cancel_token)
+                # We start with a sleep because our _run() method will setup the initial portmap.
+                await wait_with_token(
+                    asyncio.sleep(self._nat_portmap_lifetime), token=self.cancel_token)
+                await self.add_nat_portmap()
             except OperationCancelled:
                 break
 
-    async def _add_nat_portmap(self, lifetime: int) -> None:
+    async def add_nat_portmap(self) -> None:
+        self.logger.info("Setting up NAT portmap...")
+        # This is experimental and it's OK if it fails, hence the bare except.
+        try:
+            upnp_dev = await self._discover_upnp_device()
+            if upnp_dev is None:
+                return
+            await self._add_nat_portmap(upnp_dev)
+        except Exception as e:
+            if (isinstance(e, upnpclient.soap.SOAPError) and
```

Comment:
This seems like it'd be cleaner as it's own except block even if it involves having the `self.logger.exception` section in both `except` blocks.

Reply:
Sounds good to me